### DISCUSSION TITLE
Docker container's entrypoint ignored/overriden in GitHub Actions

### DISCUSSION THREAD

**Discussion by baip** [community/community#24944] [Actions]
<p>I’m trying to port a GitLab Pipeline to GitHub Actions, where we use Docker containers to provide the runtime environment. In GitLab, we simply use a line <code>image: $DOCKER_TAG</code>. The images are built by ourselves, which use a script as the entry point <code>ENTRYPOINT ["/run.sh"]</code>. The script sets up environment (e.g., by sourcing the <code>setvars.sh</code> script for the Intel compilers and calling <code>ulimit -s unlimited</code>, etc.) and calls <code>exec "$@"</code> at the end. For GitHub, I am using</p>
<pre><code class="lang-auto">container:
  image: ${{ matrix.DOCKER_TAG }}
</code></pre>
<p>However, the commands to be run later cannot find the needed binaries. Looking at the log, it appears that the container was created with <code>--entrypoint "tail"</code>, causing the <code>run.sh</code> script to be ignored. I tried adding <code>options: --entrypoint '/run.sh'</code> in the Workflow YAML file, but it did not get reflected in how the container was created and the command still failed.</p>
<p>I may be missing something obvious, though I checked both the documentation and Google. Is there any way to use the entrypoint provided by the image without creating a Docker container action?</p>

**Comment by baip** [comment]
<p>Two more things I tried:</p>
<ol>
<li>Specifying the <code>/run.sh</code> script as <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#custom-shell" rel="noopener nofollow ugc">Custom shell</a>: <code>shell: '/run.sh {0}'</code>, but got an error</li>
</ol>
<pre><code class="lang-auto">Error: Second path fragment must not be a drive or UNC name. (Parameter 'expression')
</code></pre>
<ol start="2">
<li>Using <a href="https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action" rel="noopener nofollow ugc">Docker container action</a> or <a href="https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-a-docker-hub-action" rel="noopener nofollow ugc">specifying a Docker image to use for a job step</a>. But in both cases the Docker image has to be hard coded (or built fresh every time). Trying to use input arguments like</li>
</ol>
<pre><code class="lang-auto"># Docker container action
image: docker://${{ inputs.docker_tag }}
</code></pre>
<p>or</p>
<pre><code class="lang-auto"># Job step
- uses: docker://${{ matrix.DOCKER_TAG }}
  with:
    args: ./.github/actions/build/build.sh
</code></pre>
<p>will both get an error</p>
<pre><code class="lang-auto">Unrecognized named-value: 'inputs'. Located at position 1 within expression: inputs.docker_tag
</code></pre>

**Comment by baip** [comment, marked as answer]
<p>I’ve settled along the lines below. Not ideal/DRY, as the <code>run.sh</code> entrypoint script has to be duplicated from the Docker container and kept up to date. Also, the <a href="https://github.com/actions/upload-artifact" rel="noopener nofollow ugc">upload-artifact</a> GitHub Actions does not preserve executable bits, so have to zip everything in a tar file.</p>
<pre><code class="lang-auto">jobs:
  build:
    container:
      image: XX/compiler:${{ matrix.DOCKER_TAG }}
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - run: ./.github/scripts/run.sh ./.github/scripts/build.sh
      - uses: actions/upload-artifact@v2
        with:
          name: build-artifact
          path: 'build-*.tar.bz2'
          retention-days: 7
    strategy:
      fail-fast: false
      matrix:
        DOCKER_TAG: [gcc, nvhpc, intel]
        include:
          - DOCKER_TAG: gcc
            FC: gfortran
          - DOCKER_TAG: nvhpc
            FC: nvfortran
          - DOCKER_TAG: intel
            FC: ifort
</code></pre>

**Comment by CMCDragonkai** [comment]
Why does Github CI not automatically execute the default entrypoint? It's also not really documented anywhere?
